timemachine/cpp/src/nonbonded_interaction_group.cu

```diff
@@ -0,0 +1,480 @@
+        for (int i = 0; i < 9; i++) {
+            if (i == 0 || i == 4 || i == 8) {
+                if (h_box[i] < db_cutoff) {
+                    throw std::runtime_error(
+                        "Cutoff with padding is more than half of the box width, neighborlist is no longer reliable");
+                }
+            } else if (h_box[i] != 0.0) {
+                throw std::runtime_error("Provided non-ortholinear box, unable to compute nonbonded energy");
+            }
+        }
+
+        gpuErrchk(cudaMemsetAsync(d_rebuild_nblist_, 0, sizeof(*d_rebuild_nblist_), stream));
+        gpuErrchk(cudaMemcpyAsync(d_nblist_x_, d_x, N * 3 * sizeof(*d_x), cudaMemcpyDeviceToDevice, stream));
+        gpuErrchk(cudaMemcpyAsync(d_nblist_box_, d_box, 3 * 3 * sizeof(*d_box), cudaMemcpyDeviceToDevice, stream));
+    } else {
+        k_permute<<<dimGrid, tpb, 0, stream>>>(N, d_perm_, d_x, d_sorted_x_);
+        gpuErrchk(cudaPeekAtLastError());
+    }
+
+    // do parameter interpolation here
+    if (Interpolated) {
+        CUresult result = compute_permute_interpolated_.configure(dimGrid, tpb, 0, stream)
+                              .launch(lambda, N, d_perm_, d_p, d_sorted_p_, d_sorted_dp_dl_);
+        if (result != 0) {
+            throw std::runtime_error("Driver call to k_permute_interpolated failed");
+        }
+    } else {
+        k_permute<<<dimGrid, tpb, 0, stream>>>(N, d_perm_, d_p, d_sorted_p_);
+        gpuErrchk(cudaPeekAtLastError());
+        gpuErrchk(cudaMemsetAsync(d_sorted_dp_dl_, 0, N * 3 * sizeof(*d_sorted_dp_dl_), stream))
+    }
+
+    // this stream needs to be synchronized so we can be sure that p_ixn_count_ is properly set.
+    // reset buffers and sorted accumulators
+    if (d_du_dx) {
+        gpuErrchk(cudaMemsetAsync(d_sorted_du_dx_, 0, N * 3 * sizeof(*d_sorted_du_dx_), stream))
+    }
+    if (d_du_dp) {
+        gpuErrchk(cudaMemsetAsync(d_sorted_du_dp_, 0, N * 3 * sizeof(*d_sorted_du_dp_), stream))
+    }
+
+    // update new w coordinates
+    // (tbd): cache lambda value for equilibrium calculations
+    // TODO: skip computing w coords for non-interacting atoms?
+    CUresult result = compute_w_coords_instance_.configure(B, tpb, 0, stream)
+                          .launch(N, lambda, cutoff_, d_lambda_plane_idxs_, d_lambda_offset_idxs_, d_w_, d_dw_dl_);
+    if (result != 0) {
+        throw std::runtime_error("Driver call to k_compute_w_coords");
+    }
+
+    gpuErrchk(cudaPeekAtLastError());
+    k_permute_2x<<<B, tpb, 0, stream>>>(N, d_perm_, d_w_, d_dw_dl_, d_sorted_w_, d_sorted_dw_dl_);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // look up which kernel we need for this computation
+    int kernel_idx = 0;
+    kernel_idx |= d_du_dp ? 1 << 0 : 0;
+    kernel_idx |= d_du_dl ? 1 << 1 : 0;
+    kernel_idx |= d_du_dx ? 1 << 2 : 0;
+    kernel_idx |= d_u ? 1 << 3 : 0;
+
+    if (p_ixn_count_[0] > 0) {
```

Comment:
If this is the case, can we not just return from the function immediately as nothing will change?

Reply:
> Also how often is this check false?

Probably not very often -- I hit it initially when there was a bug in the neighborlist invocation resulting in some test cases having zero interactions. I added this guard to protect against a runtime error in this case (the kernel call fails if passed 0 for number of blocks), but I think you're right that it makes sense to just return early if we hit this case.

Reply:
Added a test for the case of no interactions in d0f154a, and early return in 0cd9168